core/src/Kokkos_ExecPolicy.hpp

```diff
@@ -864,6 +864,178 @@ ThreadVectorRange(const TeamMemberType&, const iType1& arg_begin,
+  static constexpr auto par_thread = Impl::TeamMDRangeParThread::NotParThread;
+  static constexpr auto par_vector = Impl::TeamMDRangeParVector::ParVector;
+
+  static constexpr Iterate direction =
+      iter == Iterate::Default
+          ? layout_iterate_type_selector<ArrayLayout>::outer_iteration_pattern
+          : iter;
+
+  template <class... Args>
+  KOKKOS_INLINE_FUNCTION ThreadVectorMDRange(TeamHandleType const& team_,
+                                             Args&&... args)
+      : team(team_), boundaries{static_cast<BoundaryType>(args)...} {
+    static_assert(sizeof...(Args) == total_nest_level);
+  }
+
+  TeamHandleType const& team;
+  BoundaryType boundaries[total_nest_level];
+};
+
+template <typename TeamHandle, typename... Args>
+ThreadVectorMDRange(TeamHandle const&, Args&&...)
+    ->ThreadVectorMDRange<Rank<sizeof...(Args), Iterate::Default>, TeamHandle>;
+
+template <typename Rank, typename TeamHandle>
+struct TeamVectorMDRange;
+
+template <unsigned N, Iterate OuterDir, Iterate InnerDir, typename TeamHandle>
+struct TeamVectorMDRange<Rank<N, OuterDir, InnerDir>, TeamHandle> {
+  using NestLevelType  = int;
+  using BoundaryType   = int;
+  using TeamHandleType = TeamHandle;
+  using ExecutionSpace = typename TeamHandleType::execution_space;
+  using ArrayLayout    = typename ExecutionSpace::array_layout;
+
+  static constexpr NestLevelType total_nest_level =
+      Rank<N, OuterDir, InnerDir>::rank;
+  static constexpr Iterate iter    = OuterDir;
+  static constexpr auto par_thread = Impl::TeamMDRangeParThread::ParThread;
+  static constexpr auto par_vector = Impl::TeamMDRangeParVector::ParVector;
+
+  static constexpr Iterate direction =
+      iter == Iterate::Default
+          ? layout_iterate_type_selector<ArrayLayout>::outer_iteration_pattern
+          : iter;
+
+  template <class... Args>
+  KOKKOS_INLINE_FUNCTION TeamVectorMDRange(TeamHandleType const& team_,
+                                           Args&&... args)
+      : team(team_), boundaries{static_cast<BoundaryType>(args)...} {
+    static_assert(sizeof...(Args) == total_nest_level);
+  }
+
+  TeamHandleType const& team;
+  BoundaryType boundaries[total_nest_level];
+};
+
+template <typename TeamHandle, typename... Args>
+TeamVectorMDRange(TeamHandle const&, Args&&...)
+    ->TeamVectorMDRange<Rank<sizeof...(Args), Iterate::Default>, TeamHandle>;
+
+template <typename Rank, typename TeamHandle, typename Lambda,
+          typename ReducerValueType>
+KOKKOS_INLINE_FUNCTION void parallel_reduce(
+    TeamThreadMDRange<Rank, TeamHandle> const& policy, Lambda const& lambda,
+    ReducerValueType& val) {
+  Impl::md_parallel_impl<Rank>(policy, lambda, val);
+}
+
+template <typename Rank, typename TeamHandle, typename Lambda>
+KOKKOS_INLINE_FUNCTION void parallel_for(
+    TeamThreadMDRange<Rank, TeamHandle> const& policy, Lambda const& lambda) {
+  Impl::md_parallel_impl<Rank>(policy, lambda, Impl::no_reduction_tag);
```

Comment:
```suggestion
  Impl::md_parallel_impl<Rank>(policy, lambda, Impl::NoReductionTag{});
```
here and elsewhere.

Reply:
agreed lets get rid of the global static variable and just create it in-place.

Reply:
Changed in [f5f24e3](https://github.com/kokkos/kokkos/pull/5238/commits/f5f24e33bb5c604fb0cbc471850d201f86ca308c).